tty/gr716-uart/gr716-uart.c

```diff
@@ -0,0 +1,463 @@
+	ctl.type = pctl_cguctrl;
+
+	ctl.cguctrl.state = enable;
+	ctl.cguctrl.cgu = cgu_primary;
+	ctl.cguctrl.cgudev = cgudev;
+
+	return platformctl(&ctl);
+}
+
+
+static int uart_init(unsigned int n, speed_t baud, int raw)
+{
+	libtty_callbacks_t callbacks;
+	uart_t *uart = &uart_common.uart;
+
+	if (uart_setPin(info[n].rxPin) < 0 || uart_setPin(info[n].txPin) < 0 || uart_cguInit(info[n].cgudev) < 0) {
+		return -EINVAL;
+	}
+
+	uart->base = info[n].base;
+
+	callbacks.arg = uart;
+	callbacks.set_cflag = uart_setCFlag;
+	callbacks.set_baudrate = uart_setBaudrate;
+	callbacks.signal_txready = uart_signalTXReady;
+
+	if (libtty_init(&uart->tty, &callbacks, _PAGE_SIZE, baud) < 0) {
+		return -ENOENT;
+	}
+
+	if (condCreate(&uart->cond) != EOK) {
+		libtty_close(&uart->tty);
+		libtty_destroy(&uart->tty);
+		return -ENOENT;
+	}
+
+	if (mutexCreate(&uart->lock) != EOK) {
+		libtty_close(&uart->tty);
+		libtty_destroy(&uart->tty);
+		resourceDestroy(uart->cond);
+		return -ENOENT;
+	}
+
+	/* Set raw mode */
+	if (raw == 1) {
+		libtty_set_mode_raw(&uart->tty);
+	}
+
+	/* normal mode, 1 stop bit, no parity, 8 bits */
+	uart_setCFlag(uart, &uart->tty.term.c_cflag);
+
+	uart->tty.term.c_ispeed = uart->tty.term.c_ospeed = baud;
+	uart_setBaudrate(uart, baud);
+	*(uart->base + uart_ctrl) = RX_INT | RX_EN | TX_EN;
+
+	beginthread(uart_intThread, 4, &uart->stack, sizeof(uart->stack), (void *)uart);
+	interrupt(info[n].irq, uart_interrupt, uart, uart->cond, &uart->inth);
+
+	return EOK;
+}
+
+
+static void uart_usage(const char *progname)
+{
+	printf("Usage: %s [options]\n", progname);
+	printf("Options:\n");
+	printf("\t-b <baudrate>   - baudrate\n");
+	printf("\t-n <id>		  - uart controller id\n");
+	printf("\t-r              - set raw mode (default cooked)\n");
+	printf("\t-h              - print this message\n");
+}
+
+int main(int argc, char **argv)
+{
+	int uartn = UART_CONSOLE_USER;
+	speed_t baud = B230400;
+	int c, raw = 0;
+
+	if (argc > 1) {
+		while ((c = getopt(argc, argv, "n:b:rh")) != -1) {
+			switch (c) {
+				case 'b':
+					baud = libtty_int_to_baudrate(atoi(optarg));
+					if (baud == (speed_t)-1) {
+						debug("gr716-uart: wrong baudrate value\n");
+						return EXIT_FAILURE;
+					}
+					break;
+
+				case 'n':
+					uartn = atoi(optarg);
+					if (uartn >= UART_MAX_CNT) {
+						debug("gr716-uart: wrong uart ID\n");
+						return EXIT_FAILURE;
+					}
+					break;
+
+				case 'r':
+					raw = 1;
+					break;
+
+				case 'h':
+					uart_usage(argv[0]);
+					return EXIT_SUCCESS;
+
+				default:
+					uart_usage(argv[0]);
+					return EXIT_FAILURE;
+			}
+		}
+	}
+
+	if (uartn < 0) {
+		debug("gr716-uart: wrong uart ID, this uart cannot be a console\n");
+		return EXIT_FAILURE;
+	}
+
+	portCreate(&uart_common.uart.oid.port);
+
+	if (uart_init(uartn, baud, raw) < 0) {
+		debug("gr716-uart: cannot initialize uart\n");
+		return EXIT_FAILURE;
+	}
+
+	beginthread(uart_dispatchMsg, 4, uart_common.stack, sizeof(uart_common.stack), NULL);
```

Comment:
4 is a somewhat low priority for a hardware driver
```suggestion
	beginthread(uart_dispatchMsg, 3, uart_common.stack, sizeof(uart_common.stack), NULL);
```